### ISSUE TITLE
Open connections closed due to unconsumed input

### ISSUE THREAD

**Issue by treetip** [eclipse/jetty.project#4117]
# Short version

Jetty 9.4.x via Spring boot 2.x (repro project has more details), tested on JDK 11 and 8.

Jetty server unexpectedly closes connections due to "unconsumed input" causing the client to abort. But configuring a servlet filter to check the request inputstream just returns -1.

See [treetip/jetty_test](https://github.com/treetip/jetty_test) for reproduction.

# Long version

Our project has multiple micro'ish services running a combination of Jersey and Jetty via Spring Boot 2.x.

If a service calls another service it's via a Jersey Client configured with `PoolingHttpClientConnectionManager` for Apache HttpClient.

Recently we ran into a problem where

* A timed job splits work into batches and calls Service A with the batches
* Service A in turn calls Services B, C, D ... etc to enrich the data
* Service B in the chain suddenly started failing when called closing the pooled connection after a
  few requests, resulting the caller raising an exception on "Broken Pipe (Write failed)"

Even though the services dependencies are carbon copies of each other, only one of the services is failing.

Reproduction project with instructions at: [treetip/jetty_test](https://github.com/treetip/jetty_test)

Checking with wireshark a single tcp connection seems to handle a few http 1.1 requests, until Jetty closes the connection due to uncosumed input. The client then receives a RST when it tries to reuse the connection that was forcibly closed.

[Issue 2867](https://github.com/eclipse/jetty.project/issues/2867) seemed to have similar issues, so I tested the filter solution mentioned there, and it works. Though the InputStream seems to always only return -1 after control is returned to the filter (so what's actually unconsumed here?).

Changing the Spring dependency in the example project from `spring-boot-starter-jetty` to `spring-boot-starter-undertow` also fixes the problem.

Also if service A sends an empty json array, the problem goes away, which seems weird since it's still input.

So the actual question here is what should we do? Is this a bug that can be fixed? Configuring a filter that does nothing in all our services seems pointless.

I just need to understand what is happening and why, as this might reappear in other services if this is about some specific circumstances.

**Comment by joakime** [comment]
As you've discovered, "unconsumed input" is when the server component doesn't read the request body content.

That means the HTTP/1.1 connection cannot be kept open (persistent) if the entire request and response hasn't been read by both sides (client and server).
The connection must be closed in this scenario, regardless of the `Connection` header.

This is often seen with various parsing libraries ([JSON](https://github.com/eclipse/jetty.project/issues/3027#issuecomment-433379857) and XML being the main culprits).
They will read until the document is closed (the final `"}"` or `"]"` on JSON or the closing of the root element in XML).
They never read to EOF (-1).
This is a problem, and we don't know if the entire contents were read or not.
This problem is especially problematic when `Transfer-Encoding: chunked` is present, as the final chunk isn't read, leaving the chunked state open.

As far as your treetip/jetty_test sample project, it doesn't compile. (server cannot be compiled due to missing javax.servlet dependency, and server/src/main/java/test/jetty/config/Filter.java is not compilable as it doesn't use the Filter interface properly)

Why does your `/client` and `/server` tree both have a `Server.java` ??

gradle is a poor choice for reproduction projects, as it does not produce the same build for everyone, so I went ahead mavenized it.
the fork can be found at https://github.com/joakime/issue-4117-unconsumed-input

**Comment by treetip** [comment]
> As far as your treetip/jetty_test sample project, it doesn't compile. (server cannot be compiled due to missing javax.servlet dependency, and server/src/main/java/test/jetty/config/Filter.java is not compilable as it doesn't use the Filter interface properly)

Worked fine for me and a coworker I had test it using `./gradlew bootRun` to run the spring boot app.
IDEA import from existing sources as gradle project also worked fine.

This is basically how our http services are set up, spring boot gradle projects, just wanted to make sure the setup is similar to ours when making sure it's not something about our setup specifically, but a really simple project also has this.

> Why does your /client and /server tree both have a Server.java ??

Just the entrypoint for the spring boot app. Copy pasted the initial project and added the rest api that in turn calls the server/ projects api.

> As you've discovered, "unconsumed input" is when the server component doesn't read the request body content.

I guess the main question here from me is, why does it happen only in this service (we have tens of these). And specifically, how and why does the Filter help fix the problem since it just reads a -1, indicating the stream was already read?

**Comment by joakime** [comment]
The forked version is fixed to actually compile now.

Running it with wireshark and debug logging shows the following ...

The POST requests from the RestClient to the Rest Server are as follows ...

```
POST /v1/test HTTP/1.1
Accept: application/json
User-Agent: Jersey/2.27 (Apache HttpClient 4.5)
Content-Type: application/json
Transfer-Encoding: chunked
Host: localhost:6700
Connection: Keep-Alive
Accept-Encoding: gzip,deflate
```

And when the server issues the `unconsumed input` is when the chunked request body content is not read fully to EOF (-1), leaving an unconsumed chunked transfer in-progress and unclosed.

**Comment by sbordet** [comment]
>  And specifically, how and why does the Filter help fix the problem since it just reads a -1, indicating the stream was already read?

The filter issues a read that Jetty performs on the network. Jetty reads the terminal chunk `0\r\n\r\n` that is 5 bytes and now it knows that the chunked content is really finished. It could be more bytes if there are HTTP trailers.

Jetty does not give to the application the 5 bytes terminal chunk, but just -1 to indicate that now the content is finished.

So "just reads a -1" is more complicated than typically expected.
If the application does not issue this last read, Jetty does not parse the terminal chunk and does not know whether the content is really finished.

**Comment by joakime** [comment]
One final commit to the forked example project fixes your jackson issue.

See https://github.com/joakime/issue-4117-unconsumed-input/commit/1823e2d00e179d714d0c36e462776c9cb740808a

This was discovered in https://github.com/FasterXML/jackson-jaxrs-providers/issues/108
And was detailed in https://medium.com/@cowtowncoder/jackson-2-9-features-b2a19029e9ff (search "No Trailing Junk")

This is not a Jetty bug or issue.

**Comment by treetip** [comment]
Nice, thank you.

Somehow my searching around did not stumble on that.

**Comment by stevenschlansker** [comment]
Hi @joakime @sbordet , sorry to raise a dead thread, but wanted to let you know that [this issue](https://github.com/FasterXML/jackson-jaxrs-providers/issues/108) should be fixed upstream by default in Jackson 2.15. So hopefully you will no longer have to re-diagnose this issue on Jackson's behalf going forward :)
